doc/psyke.rst

```diff
@@ -0,0 +1,287 @@
+
+* The entire Invoke (extraction applied to all Nodes).
+
+The basic mechanism of code extraction is applying the **ExtractRegionTrans**
+transformation to selected Nodes. This transformation is further sub-classed
+into API-specific implementations, **DynamoExtractRegionTrans** and
+**GOceanExtractRegionTrans**. Both sub-classed transformations insert an
+instance of the **ExtractNode** object into the Schedule of a specific Invoke.
+All Nodes marked for extraction become children of the **ExtractNode**.
+
+For now, the **ExtractNode** class simply adds comments at the beginning
+and the end of the extract region, that is at the position of the extract
+Node and after all its children. For example, marking a single Loop
+containing a Kernel call in Dynamo0.3 API generates:
+
+.. code-block:: fortran
+
+      ! ExtractStart
+      ! CALL write_extract_arguments(argument_list)
+      !
+      DO cell=1,f3_proxy%vspace%get_ncell()
+        !
+        CALL testkern_code_w2_only(nlayers, f3_proxy%data, f2_proxy%data, ...)
+      END DO 
+      !
+      ! ExtractEnd
+
+The ``! CALL write_extract_arguments(argument_list)`` comment will be
+replaced by appropriate calls to write out arguments of extracted Node(s)
+or Kernel(s) in Issue #234. The **ExtractNode** base class can be sub-classed
+by API-specific code. This will be required for constructing the list of
+quantities required by Kernel (and Built-In) calls in the extracted Nodes.
+
+.. _psyke-intro-restrictions:
+
+Restrictions
+++++++++++++
+
+Code extraction can be applied to unoptimised or optimised code. There are
+restrictions that check for correctness of optimising transformations when
+extraction is applied, as well as restrictions that eliminate dependence on
+the specific model infrastructure.
+
+.. _psyke-intro-restrictions-gen:
+
+General
+#######
+
+This group of restrictions is enforced irrespective of whether optimisations
+are used or not.
+
+* Extraction can be applied to a single Node in or a list of Nodes in a
+  Schedule. For the latter, Nodes in the list must be consecutive children
+  of the same parent Schedule.
+
+* Extraction cannot be applied to an **ExtractNode** or a Node list that
+  already contains one (otherwise we would have an extract region within
+  another extract region).
+
+* A Kernel or a Built-In call cannot be extracted without its parent Loop.
+
+.. _psyke-intro-restrictions-dm:
+
+Distributed memory
+##################
+
+As noted in the :ref:`distributed_memory` section, support for distributed
+memory in PSyclone is currently limited to the Dynamo0.3 API. Since the
+implementation generates calls to LFRic infrastructure (e.g. runtime checks
+for status of field halos), code extraction when distributed memory is
+enabled is not allowed.
+
+.. _psyke-intro-restrictions-shared:
+
+Shared memory and API-specific
+##############################
+
+The restrictions below do not allow the **ExtractRegionTrans** to be
```

Comment:
Suggest: "The **ExtractRegionTrans** transformation cannot be applied to:"

Reply:
Definitely simpler. Done.